gokart/redis_lock.py

```diff
@@ -80,6 +92,80 @@ def extend_lock():
     return wrapper
 
 
+def wrap_with_run_lock(func, redis_params: RedisParams):
+    """Redis lock wrapper function for RunWithLock.
+    When a fucntion is wrapped by RunWithLock, the wrapped function will be simply wrapped with redis lock.
+    https://github.com/m3dev/gokart/issues/265
+    """
+    return _wrap_with_lock(func=func, redis_params=redis_params)
+
+
+def wrap_with_dump_lock(func: Callable, redis_params: RedisParams, exist_check: Callable):
+    """Redis lock wrapper function for TargetOnKart.dump().
+    When TargetOnKart.dump() is called, dump() will be wrapped with redis lock and cache existance check.
+    https://github.com/m3dev/gokart/issues/265
+    """
+
+    if not redis_params.should_redis_lock:
+        return func
+
+    def wrapper(*args, **kwargs):
+        redis_lock = _set_redis_lock(redis_params=redis_params)
+        scheduler = _set_lock_scheduler(redis_lock=redis_lock, redis_params=redis_params)
+
+        try:
+            logger.debug(f'Task lock of {redis_params.redis_key} locked.')
+            if not exist_check():
+                func(*args, **kwargs)
+            redis_lock.release()
+            logger.debug(f'Task lock of {redis_params.redis_key} released.')
+            scheduler.shutdown()
+        except BaseException as e:
+            logger.debug(f'Task lock of {redis_params.redis_key} released with BaseException.')
+            redis_lock.release()
+            scheduler.shutdown()
+            raise e
+
+    return wrapper
+
+
+def wrap_with_load_lock(func, redis_params: RedisParams):
+    """Redis lock wrapper function for TargetOnKart.load().
+    When TargetOnKart.load() is called, redis lock will be locked and released before load().
+    https://github.com/m3dev/gokart/issues/265
+    """
+
+    if not redis_params.should_redis_lock:
+        return func
+
+    def wrapper(*args, **kwargs):
+        redis_lock = _set_redis_lock(redis_params=redis_params)
+        scheduler = _set_lock_scheduler(redis_lock=redis_lock, redis_params=redis_params)
+
+        try:
+            logger.debug(f'Task lock of {redis_params.redis_key} locked.')
+            redis_lock.release()
+            logger.debug(f'Task lock of {redis_params.redis_key} released.')
+            scheduler.shutdown()
+            result = func(*args, **kwargs)
+            return result
+        except BaseException as e:
```

Comment:
Where do you think BaseException will occur from?

* If `redis_lock.release()` or `scheduler.shutdown()` in `try` block raise Exception, the same function in `except` block will occur again and overwrap the exception which might be other exception.
* If `func(*args, **kwargs)` raise Exception, it doesn't need to unlock because unlock method already called.

So, I think try block is unnecessary in this function.

Reply:
@kitagry Thank you for checking carefully! 

This try block was for caching Exception in func(). Therefore, as you mentioned, this try block seems to be unnecessary. I will delete this try block!